File: Simplified/NYPLBookState.m
//#import "NYPLBookState.h"
//
//static NSString *const Downloading = @"downloading";
//static NSString *const DownloadFailed = @"download-failed";
//static NSString *const DownloadNeeded = @"download-needed";
//static NSString *const DownloadSuccessful = @"download-successful";
//static NSString *const Unregistered = @"unregistered";
//static NSString *const Holding = @"holding";
//static NSString *const Used = @"used";
//static NSString *const Unsupported = @"unsupported";
//
//NYPLBookState NYPLBookStateFromString(NSString *const string)
//{
//  if([string isEqualToString:Downloading]) return NYPLBookStateDownloading;
//  if([string isEqualToString:DownloadFailed]) return NYPLBookStateDownloadFailed;
//  if([string isEqualToString:DownloadNeeded]) return NYPLBookStateDownloadNeeded;
//  if([string isEqualToString:DownloadSuccessful]) return NYPLBookStateDownloadSuccessful;
//  if([string isEqualToString:Unregistered]) return NYPLBookStateUnregistered;
//  if([string isEqualToString:Holding]) return NYPLBookStateHolding;
//  if([string isEqualToString:Used]) return NYPLBookStateUsed;
//  if([string isEqualToString:Unsupported]) return NYPLBookStateUnsupported;
//
//  @throw NSInvalidArgumentException;
//}
//
//NSString *NYPLBookStateToString(NYPLBookState state)
//{
//  switch(state) {
//    case NYPLBookStateDownloading:
//      return Downloading;
//    case NYPLBookStateDownloadFailed:
//      return DownloadFailed;
//    case NYPLBookStateDownloadNeeded:
//      return DownloadNeeded;
//    case NYPLBookStateDownloadSuccessful:
//      return DownloadSuccessful;
//    case NYPLBookStateUnregistered:
//      return Unregistered;
//    case NYPLBookStateHolding:
//      return Holding;
//    case NYPLBookStateUsed:
//      return Used;
//    case NYPLBookStateUnsupported:
//      return Unsupported;
//  }
//}

File: Simplified/NYPLBookState.swift
import Foundation

let DownloadingKey = "downloading"
let DownloadFailedKey = "download-failed"
let DownloadNeededKey = "download-needed"
let DownloadSuccessfulKey = "download-successful"
let UnregisteredKey = "unregistered"
let HoldingKey = "holding"
let UsedKey = "used"
let UnsupportedKey = "unsupported"

@objc public enum NYPLBookState : Int {
  case Unregistered = 0
  case DownloadNeeded = 1
  case Downloading
  case DownloadFailed
  case DownloadSuccessful
  case Holding
  case Used
  case Unsupported
    
  init?(_ stringValue: String) {
    switch stringValue {
      case DownloadingKey:
        self = .Downloading
      case DownloadFailedKey:
        self = .DownloadFailed
      case DownloadNeededKey:
        self = .DownloadNeeded
      case DownloadSuccessfulKey:
        self = .DownloadSuccessful
      case UnregisteredKey:
        self = .Unregistered
      case HoldingKey:
        self = .Holding
      case UsedKey:
        self = .Used
      case UnsupportedKey:
        self = .Unsupported
      default:
        return nil
    }
  }
    
  func stringValue() -> String {
    switch self {
      case .Downloading:
        return DownloadingKey;
      case .DownloadFailed:
        return DownloadFailedKey;
      case .DownloadNeeded:
        return DownloadNeededKey;
      case .DownloadSuccessful:
        return DownloadSuccessfulKey;
      case .Unregistered:
        return UnregisteredKey;
      case .Holding:
        return HoldingKey;
      case .Used:
        return UsedKey;
      case .Unsupported:
        return UnsupportedKey;
    }
  }
}

// For Objective-C, since Obj-C enum is not allowed to have methods
// TODO: Remove when migration to Swift completed
class NYPLBookStateHelper : NSObject {
  @objc(stringValueFromBookState:)
  static func stringValue(from state: NYPLBookState) -> String {
    return state.stringValue()
  }
    
  @objc(bookStateFromString:)
  static func bookState(fromString string: String) -> Int {
    return NYPLBookState.init(string)?.rawValue ?? -1
  }
    
  @objc static func allBookStates() -> [NYPLBookState.RawValue] {
    return [
      NYPLBookState.Unregistered.rawValue,
      NYPLBookState.DownloadNeeded.rawValue,
      NYPLBookState.Downloading.rawValue,
      NYPLBookState.DownloadFailed.rawValue,
      NYPLBookState.DownloadSuccessful.rawValue,
      NYPLBookState.Holding.rawValue,
      NYPLBookState.Used.rawValue,
      NYPLBookState.Unsupported.rawValue
    ]
  }
}

